apps/prairielearn/src/middlewares/authzHasCoursePreviewOrInstanceView.tsx

```diff
@@ -31,28 +37,36 @@ export async function authzHasCoursePreviewOrInstanceView(req: Request, res: Res
     // If a CSRF token is not present, we fall through to the error below.
     res.locals.__csrf_token
   ) {
+    // Try to redirect to an accessible page. If we can't, then show the error page.
+    const redirectUrl = getRedirectForEffectiveAccessDenied(res);
+    if (redirectUrl) {
+      return { type: 'redirect', value: redirectUrl };
+    }
```

Comment:
We've used `throw new HttpRedirect(...)` elsewhere, which would save us from having to change callers here. I understand that using an exception here might feel weird though 🙂